atomicfu/src/commonMain/kotlin/kotlinx/atomicfu/locks/Mutex.kt

```diff
@@ -0,0 +1,20 @@
+package kotlinx.atomicfu.locks
+
+/**
+ * Multiplatform mutex.
+ * On native based on futex(-like) system calls.
+ * On JVM delegates to ReentrantLock.
+ */
+expect class Mutex {
+    fun isLocked(): Boolean
+    fun tryLock(): Boolean
+    fun lock()
+    fun unlock()
+}
+
+fun <T> Mutex.withLock(block: () -> T): T {
+    lock()
+    val result = block()
+    unlock()
+    return result
+}
```

Comment:
`kotlinx.coroutines` could provide a couple of useful insights. For example, the `withLock` of asynchronous locks (https://github.com/Kotlin/kotlinx.coroutines/blob/6c6df2b850382887462eeaf51f21f58bd982491d/kotlinx-coroutines-core/common/src/sync/Mutex.kt#L117-L127):

Ensures that `unlock()` is called even if `block()` throws an exception:

```kotlin
val mutex = Mutex()
try {
  mutex.withLock {
    throw IllegalStateException()
  }
} catch (e: IllegalStateException) { }
println(mutex.isLocked())
```

Allows non-local returns:

```kotlin
val mutex = Mutex()
fun something() = mutex.withLock {
  if (3 < 4) return
  println("3 >= 4?!!")
}
```

Specifies via a contract that the block is being executed exactly once:

```kotlin
val mutex = Mutex()
val i: Int
mutex.withLock {
  i = 3 // late initialization, is valid because the block is invoked exactly once
}
println(i)
```